src/sentry/workflow_engine/processors/action.py

```diff
@@ -127,11 +130,122 @@ def filter_recently_fired_workflow_actions(
     actions_without_statuses_ids = {action.id for action in actions_without_statuses}
     filtered_actions = actions.filter(id__in=actions_to_include | actions_without_statuses_ids)
 
+    # dual write to WorkflowActionGroupStatus
+    filter_recently_fired_workflow_actions(filtered_action_groups, event_data)
+
     create_workflow_fire_histories(filtered_actions, event_data)
 
     return filtered_actions
 
 
+def get_workflow_group_action_statuses(
+    action_to_workflows_ids: dict[int, set[int]], group: Group
+) -> dict[int, list[WorkflowActionGroupStatus]]:
+    all_statuses = WorkflowActionGroupStatus.objects.filter(
+        group=group, action_id__in=action_to_workflows_ids.keys()
+    )
+
+    actions_with_statuses: dict[int, list[WorkflowActionGroupStatus]] = defaultdict(list)
+
+    for status in all_statuses:
+        workflow_id = status.workflow_id
+        action_id = status.action_id
+        if workflow_id not in action_to_workflows_ids.get(action_id, []):
+            # if the (workflow, action) combination shouldn't be processed, skip it
+            # more difficult to query than to iterate
+            continue
+
+        actions_with_statuses[action_id].append(status)
+
+    return actions_with_statuses
+
+
+def update_workflow_action_group_statuses(
+    action_to_workflows_ids: dict[int, set[int]],
+    action_to_statuses: dict[int, list[WorkflowActionGroupStatus]],
+    workflows: BaseQuerySet[Workflow],
+    group: Group,
+) -> set[int]:
+    now = timezone.now()
+    status_ids: set[int] = set()
+    workflow_frequencies = {
+        workflow.id: workflow.config.get("frequency", 0) * timedelta(minutes=1)
+        for workflow in workflows
+    }
+
+    for action_id, statuses in action_to_statuses.items():
+        for status in statuses:
+            if (now - status.date_updated) >= workflow_frequencies.get(status.workflow_id, 0):
+                # we should fire the workflow for this action
+                status_ids.add(status.id)
+
+    workflow_action_statuses = WorkflowActionGroupStatus.objects.filter(
+        id__in=status_ids, date_updated__lt=now
+    )
+    action_ids = {status.action_id for status in workflow_action_statuses}
```

Comment:
using a set will dedupe actions if multiple workflows would fire the same action. in a follow up we should know which workflow is firing it (out of all the workflows that could have done so here)